src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -258,97 +296,162 @@ void q_omg_log_exception(const struct ddsrt_log_cfg *lc, uint32_t cat, DDS_Secur
+static void remove_pending_match(struct pending_match_index *index, struct pending_match *match)
 {
-  struct guid_pair guids;
-
-  guids.src = *src;
-  guids.dst = *dst;
+  struct pending_match *found;
+  ddsrt_avl_dpath_t dpath;
 
-  return ddsrt_avl_lookup(&entity_match_treedef, &list->matches, &guids);
+  ddsrt_mutex_lock(&index->lock);
+  if ((found = ddsrt_avl_lookup_dpath(&pending_match_index_treedef, &index->pending_matches, &match->guids, &dpath)) != NULL)
+  {
+    ddsrt_avl_delete_dpath(&pending_match_index_treedef, &index->pending_matches, found, &dpath);
+    free_pending_match(found);
+  }
+  ddsrt_mutex_unlock(&index->lock);
 }
 
-static struct security_entity_match * find_or_create_entity_match(struct dds_security_match_index *list, const ddsi_guid_t *src, const ddsi_guid_t *dst)
+static void pending_match_expiry_cb(struct xevent *xev, void *varg, ddsrt_mtime_t tnow)
 {
-  struct security_entity_match *match;
+  struct pending_match_index *index = varg;
 
-  ddsrt_mutex_lock(&list->lock);
-  match = find_entity_match_locked(list, src, dst);
-  if (!match)
+  ddsrt_mutex_lock(&index->lock);
+  struct pending_match *match = ddsrt_fibheap_min(&pending_match_expiry_fhdef, &index->expiry_timers);
+  while (match && match->expiry.v <= tnow.v)
   {
-    match = entity_match_new(src, dst);
-    ddsrt_avl_insert(&entity_match_treedef, &list->matches, match);
+    ddsrt_fibheap_delete(&pending_match_expiry_fhdef, &index->expiry_timers, match);
+    ddsrt_avl_delete(&pending_match_index_treedef, &index->pending_matches, match);
+    clear_and_free_pending_match(index->gv->security_context, match);
+    match = ddsrt_fibheap_min(&pending_match_expiry_fhdef, &index->expiry_timers);
   }
-  ddsrt_mutex_unlock(&list->lock);
-
-  return match;
+  if (match)
+    resched_xevent_if_earlier(xev, match->expiry);
+  ddsrt_mutex_unlock(&index->lock);
 }
 
-static struct security_entity_match * remove_entity_match(struct dds_security_match_index *list, const ddsi_guid_t *src, const ddsi_guid_t *dst)
+static void clear_pending_matches_by_local_guid(dds_security_context *sc, struct pending_match_index *index, const ddsi_guid_t *local_guid)
 {
-  struct security_entity_match *match;
-  struct guid_pair guids;
-  ddsrt_avl_dpath_t path;
-
-  guids.src = *src;
-  guids.dst = *dst;
+  struct pending_match *match;
 
-  ddsrt_mutex_lock(&list->lock);
-  match = ddsrt_avl_lookup_dpath(&entity_match_treedef, &list->matches, &guids, &path);
-  if (match)
-    ddsrt_avl_delete_dpath(&entity_match_treedef, &list->matches, match, &path);
-  ddsrt_mutex_unlock(&list->lock);
-
-  return match;
+  ddsrt_mutex_lock(&index->lock);
+  match = ddsrt_avl_find_min(&pending_match_index_treedef, &index->pending_matches);
+  while (match)
+  {
+    struct pending_match *next = ddsrt_avl_find_succ(&pending_match_index_treedef, &index->pending_matches, match);
```

Comment:
I'm not sure whether the benefit would be worth the trouble, but you could do better than scanning all entries, you can always skip forward to the next entry that with a greater remote guid and a local guid greater-than-or-equal to.

Reply:
Implemented you suggestion. Although I would expect that the pending_match_index is mostly empty.